core/src/test/java/org/apache/iceberg/util/TestPartitionMap.java

```diff
@@ -0,0 +1,222 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.TestHelpers.Row;
+import org.apache.iceberg.data.GenericRecord;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.types.Types;
+import org.junit.jupiter.api.Test;
+
+public class TestPartitionMap {
+
+  private static final Schema SCHEMA =
+      new Schema(
+          required(1, "id", Types.IntegerType.get()),
+          required(2, "data", Types.StringType.get()),
+          required(3, "category", Types.StringType.get()));
+  private static final PartitionSpec UNPARTITIONED_SPEC = PartitionSpec.unpartitioned();
+  private static final PartitionSpec BY_DATA_SPEC =
+      PartitionSpec.builderFor(SCHEMA).identity("data").withSpecId(1).build();
+  private static final PartitionSpec BY_DATA_CATEGORY_BUCKET_SPEC =
+      PartitionSpec.builderFor(SCHEMA).identity("data").bucket("category", 8).withSpecId(3).build();
+  private static final Map<Integer, PartitionSpec> SPECS =
+      ImmutableMap.of(
+          UNPARTITIONED_SPEC.specId(),
+          UNPARTITIONED_SPEC,
+          BY_DATA_SPEC.specId(),
+          BY_DATA_SPEC,
+          BY_DATA_CATEGORY_BUCKET_SPEC.specId(),
+          BY_DATA_CATEGORY_BUCKET_SPEC);
+
+  @Test
+  public void testEmptyMap() {
+    PartitionMap<String> map = PartitionMap.create(SPECS);
+    assertThat(map).isEmpty();
+    assertThat(map).hasSize(0);
+    assertThat(map).doesNotContainKey(Pair.of(1, Row.of(1))).doesNotContainValue("value");
+    assertThat(map.values()).isEmpty();
+    assertThat(map.keySet()).isEmpty();
+    assertThat(map.entrySet()).isEmpty();
+  }
+
+  @Test
+  public void testSize() {
+    PartitionMap<String> map = PartitionMap.create(SPECS);
+    map.put(UNPARTITIONED_SPEC.specId(), null, "v1");
+    map.put(BY_DATA_SPEC.specId(), Row.of("aaa"), "v2");
+    map.put(BY_DATA_SPEC.specId(), Row.of("bbb"), "v3");
+    map.put(BY_DATA_CATEGORY_BUCKET_SPEC.specId(), Row.of("ccc", 2), "v4");
+    assertThat(map).isNotEmpty();
+    assertThat(map).hasSize(4);
+  }
+
+  @Test
+  public void testDifferentStructLikeImplementations() {
+    PartitionMap<String> map = PartitionMap.create(SPECS);
+    map.put(BY_DATA_SPEC.specId(), struct(BY_DATA_SPEC, "data", "aaa"), "value");
+    assertThat(map)
+        .containsEntry(Pair.of(BY_DATA_SPEC.specId(), struct(BY_DATA_SPEC, "data", "aaa")), "value")
+        .containsEntry(Pair.of(BY_DATA_SPEC.specId(), Row.of("aaa")), "value");
```

Comment:
This might be another good spot to do an "equals" check with two maps with the same specs but different structlikes with the same data?